policies/policy_service.go

```diff
@@ -420,4 +422,59 @@ func (s policiesService) DeleteAgentGroupFromAllDatasets(ctx context.Context, gr
 	}
 
 	return nil
+}
+
+func (s policiesService) DuplicatePolicy(ctx context.Context, token string, policyID string) (Policy, error) {
+
+	mfOwnerID, err := s.identify(token)
+	if err != nil {
+		return Policy{}, err
+	}
+
+	existingPolicy, err := s.repo.RetrievePolicyByID(ctx, policyID, mfOwnerID)
+	if err != nil {
+		return Policy{}, err
+	}
+
+	policyList, err := s.repo.RetrieveAllPoliciesInternal(ctx, mfOwnerID)
```

Comment:
Two issues: one, I'm not sure it's the best strategy to get all policies and try to figure out if the name exists or not. First, seems like a lot of work (what if there are thousand policies?) and second, this code can never be sure there isn't a race and a policy with the same name isn't inserted between the time RetrieveAllPoliciesInternal and SavePolicy is called, so it could still fail.

Even if we use this strategy, don't see why we need RetrieveAllPoliciesInternal since RetrievalAll seems to do what we want already.

I think an alternative strategy would be to not check all policies, and just make three attempts to generate a name: "OLDNAME-Copy", "OLDNAME-Copy2" and "OLDNAME-Copy3". It can try to SavePolicy the first, if it fails try the second, if it fails try the third, if it fails then fail the call.

Reply:
This was solved in the latter commit.